src/utils/prometheus.ts

```diff
@@ -0,0 +1,185 @@
+import { Counter, Gauge, Histogram, Registry } from "prom-client";
+import { getUsedBackendWallets, inspectNonce } from "../db/wallets/walletNonce";
+import { getLastUsedOnchainNonce } from "../server/routes/admin/nonces";
+
+const nonceMetrics = new Gauge({
+  name: "engine_nonces",
+  help: "Current nonce values and health for backend wallets",
+  labelNames: ["wallet_address", "chain_id", "nonce_type"],
+  async collect() {
+    const allWallets = await getUsedBackendWallets();
+
+    for (const { chainId, walletAddress } of allWallets) {
+      const [onchainNonce, engineNonce] = await Promise.all([
+        getLastUsedOnchainNonce(chainId, walletAddress),
+        inspectNonce(chainId, walletAddress),
+      ]);
+
+      this.set(
+        {
+          wallet_address: walletAddress,
+          chain_id: chainId.toString(),
+          nonce_type: "onchain",
+        },
+        onchainNonce,
+      );
+      this.set(
+        {
+          wallet_address: walletAddress,
+          chain_id: chainId.toString(),
+          nonce_type: "engine",
+        },
+        engineNonce,
+      );
+    }
+  },
+});
+
+export const enginePromRegister = new Registry();
+
+enginePromRegister.registerMetric(nonceMetrics);
+
+// Define strict types for each event
+type ResponseSentParams = {
+  endpoint: string;
+  statusCode: string;
+  method: string;
+  duration: number;
+};
+
+type TransactionQueuedParams = {
+  chainId: string;
+  walletAddress: string;
+};
+
+type TransactionSentParams = TransactionQueuedParams & {
+  success: boolean;
+  duration: number;
+};
+
+type TransactionMinedParams = TransactionQueuedParams & {
+  endToEndDuration: number;
+  duration: number;
```

Comment:
Nit: for durations can we either include the units in the name or a TSDoc comment? It was unclear without digging a few layers deep.

Also comments would be helpful since `duration` and `endToEndDuration` sound similar (if we want to keep the latter).